cache.py

```diff
@@ -0,0 +1,19 @@
+from weakref import WeakValueDictionary
+
+from entities import *
+from utils import *
+
+
+class Cache:
+    def __init__(self):
+        self._cache = WeakValueDictionary()
+
+    def add_or_update(self, transaction):
+        self._cache[transaction.header_id] = transaction
```

Comment:
Mentioned in chat, but would be easy to make this more generic just by renaming `transaction` to `object` (though `object` is a keyword in python, so maybe `entity` or `instance`) and `header_id` to `key` or `key_id` or even just `id`.